homeassistant/components/device_tracker/icloud.py

```diff
@@ -1,100 +1,449 @@
+                    self._verification_code = None
+        else:
+            self.api.authenticate()
+
+        currentminutes = dt_util.now().hour * 60 + dt_util.now().minute
+        for devicename in self.devices:
+            interval = self._intervals.get(devicename, 1)
+            if ((currentminutes % interval == 0) or
+                    (interval > 10 and
+                     currentminutes % interval in [2, 4])):
+                self.update_device(devicename)
+
+    def determine_interval(self, devicename, latitude, longitude, battery,
+                           gmtt_duration):
+        """Calculate new interval."""
+        # pylint: disable=too-many-branches,too-many-arguments
+        distancefromhome = None
+        zone_state = self.hass.states.get('zone.home')
+        zone_state_lat = zone_state.attributes['latitude']
+        zone_state_long = zone_state.attributes['longitude']
+        distancefromhome = distance(latitude, longitude, zone_state_lat,
+                                    zone_state_long)
+        distancefromhome = round(distancefromhome / 1000, 1)
+
+        currentzone = active_zone(self.hass, latitude, longitude)
+
+        if ((currentzone is not None and
+             currentzone == self._overridestates.get(devicename)) or
+                (currentzone is None and
+                 self._overridestates.get(devicename) == 'away')):
+            return
+
+        self._overridestates[devicename] = None
+
+        if currentzone is not None:
+            self._intervals[devicename] = 30
+            return
+
+        if distancefromhome is None:
+            return
+        if distancefromhome > 100:
+            self._intervals[devicename] = distancefromhome
+            if gmtt_duration is not None:
+                self._intervals[devicename] = float(gmtt_duration) - 10
+            self._intervals[devicename] = round(self._intervals[devicename], 0)
+        elif distancefromhome > 50:
+            self._intervals[devicename] = 30
+        elif distancefromhome > 25:
+            self._intervals[devicename] = 15
+        elif distancefromhome > 10:
+            self._intervals[devicename] = 5
+        else:
+            self._intervals[devicename] = 1
+        if battery is not None and battery <= 33 and distancefromhome > 3:
+            self._intervals[devicename] = self._intervals[devicename] * 2
+
+    def update_device(self, devicename):
+        """Update the device_tracker entity."""
+        # pylint: disable=too-many-branches,too-many-statements
+        # pylint: disable=too-many-nested-blocks,too-many-locals
+        from pyicloud.exceptions import PyiCloudNoDevicesException
+
+        entity = self.hass.states.get(ENTITY_ID_FORMAT.format(devicename))
+        if entity is None and devicename in self.seen_devices:
+            return
+        attrs = {}
+        kwargs = {}
+        try:
+            gtt = self.googletraveltime.get(devicename)
```

Comment:
Why do I need to remove this? It's optional and it gives this component a better result?